Various minor fixes in preparation for fixing Morello setbounds bug

arichardson opened the pull request:
This also enables the random_inputs_test for Morello by implementing the require sail wrapper functions.

arichardson committed:
Add printing of capability values in unit tests

We need to provide ostream overloads so that catch2 can print values.

arichardson committed:
Introduce a set_addr helper

This will be useful for the fuzzing tests and should be helpful for
users of the library as well.

arichardson committed:
Fix fuzzing build for RV64

Commit bc17a9a36322c5a8e76226b0f2f1293f494dbda3 changed this function to
incorrectly take a cc128_cap_t instead of the generic _cc_cap_t.

arichardson committed:
Fix random_inputs_test for Morello

We have to XOR the pesbt field with the Morello NULL mask for the
compression library but we don't do so for the sail. This resulted in
an inconsistency when when converting to/from cc128m_cap_t.
Additionally, the existing XOR used the wrong XOR mask (it was using the
RISC-V one instead of the Morello one).

arichardson committed:
Fix Morello sail wrapper to also set cr_exp

And add this check to random_inputs_test.

arichardson committed:
Re-use cap_bounds_address in compute_base_top

Also use it unconditionally instead of inside a CC_IS_MORELLO
block as it will expand to a no-op with an all-ones mask.

arichardson committed:
Reduce code duplication in get_alignment_mask()

We can reuse make_max_perms_cap here (which I believe did not exist
when get_alignment_mask was added).

arichardson committed:
Add a test case that make_max_perms_cap() matches sail

arichardson committed:
Enable NULL encoding test for Morello

The required sail APIs have now been implemented.

arichardson committed:
Set cr_exp in make_max_perms_cap

This was not caught by the existing checks since operator== does not
compare all fields.

arichardson committed:
Fully initialize the output in decompress_raw

I noticed that some tests were failing after changing operator== to
compare all fields. It turns out cr_extra was uninitialized so this
sometimes caused test failures. Fix this by setting all fields to zero
before decompressing.

arichardson committed:
Make test capability debug output more compact for test failures

arichardson committed:
Compare all fields in operator==

This makes the tests less fragile by ensuring all values have been
initialized correctly.

jrtc27 commented on cheri_compressed_cap_common.h:
```diff
@@ -337,6 +337,18 @@ static inline bool _cc_N(bounds_bits_valid)(_cc_bounds_bits bounds) {
 
+#ifdef CC_IS_MORELLO
+/// Returns the address with Morello flags (high address bits) removed and sign extended.
+static inline _cc_addr_t _cc_N(morello_cap_bounds_address)(_cc_addr_t addr) {
```
The morello_ feels redundant when it'll be cc128m_...?

arichardson replied:
True, I could just use that instead of the _cc_N macro

arichardson replied:
Dropped the commit instead.

jrtc27 commented on cheri_compressed_cap_common.h:
```diff
@@ -346,12 +358,7 @@ static inline bool _cc_N(compute_base_top)(_cc_bounds_bits bounds, _cc_addr_t cu
-    // Sign extend
-    if (cursor & ((_CC_CURSOR_MASK >> 1) + 1))
-        cursor |= ~_CC_CURSOR_MASK;
+    cursor = _cc_N(morello_cap_bounds_address)(cursor);
```
If _CC_CURSOR_MASK is ~0 on !cc128m then you can just do this unconditionally, with it becoming a slightly stupid identity

arichardson replied:
Done, thanks for the suggestion.

jrtc27 commented on cheri_compressed_cap_common.h:
```diff
@@ -723,40 +732,27 @@ static inline bool _cc_N(is_representable_new_addr)(bool sealed, _cc_addr_t base
-    return cursor;
-}
-
+#ifdef CC_IS_MORELLO
```
I'd much rather see less #ifdef'ing than more; the fact we have CC_IS_MORELLO at all rather than nicer parameterisations makes me a little sad

arichardson replied:
Dropped this.

arichardson committed:
Assert that cr_exp is valid in cap_bounds_uses_value

While touching this function also add a doc comment since the usage is
not obvious without looking at CapBoundsUsesValue from the Morello spec.
